### ISSUE TITLE
config_set_default :buffer_chunk_limit for BufferedOutput is not working

### ISSUE THREAD

**Issue by sonots** [fluent/fluentd#815]
reported by @joker1007.

TimeSlicedOutput looks as follows:

```ruby
  class TimeSlicedOutput < BufferedOutput
    ...
    config_set_default :buffer_type, 'file'  # overwrite default buffer_type
    config_set_default :buffer_chunk_limit, 256*1024*1024  # overwrite default buffer_chunk_limit
    config_set_default :flush_interval, nil

    def configure(conf)
      super
```

It tries to overwrite default `buffer_chunk_limit`, but BufferedOutput looks as follows:

```ruby
  class BufferedOutput < Output
    desc 'The buffer type (memory, file)'
    config_param :buffer_type, :string, :default => 'memory'
    desc 'The interval between data flushes.'
    config_param :flush_interval, :time, :default => 60
    config_param :try_flush_interval, :float, :default => 1
    desc 'If true, the value of `retry_value` is ignored and there is no limit'
    config_param :disable_retry_limit, :bool, :default => false
    desc 'The limit on the number of retries before buffered data is discarded'
    config_param :retry_limit, :integer, :default => 17
    desc 'The initial intervals between write retries.'
    config_param :retry_wait, :time, :default => 1.0
    desc 'The maximum intervals between write retries.'
    config_param :max_retry_wait, :time, :default => nil
    desc 'The number of threads to flush the buffer.'
    config_param :num_threads, :integer, :default => 1
    desc 'The interval between data flushes for queued chunk.'
    config_param :queued_chunk_flush_interval, :time, :default => 1

    def configure(conf)
      super

      @retry_wait = @retry_wait.to_f # converted to Float for calc_retry_wait
      @buffer = Plugin.new_buffer(@buffer_type)
      @buffer.configure(conf)
```

`config_set_default` does not change values of `conf`. And, since `buffer_chunk_limit` is an option of **Buffer** plugin, not BufferedOutput plugin, changing buffer_chunk_limit by `config_set_default` is not working for TimeSlicedOutput plugin.

FYI: Some of other plugins are imitating this way of TimeSlicedOutput and they are also not working https://github.com/kaizenplatform/fluent-plugin-bigquery/blob/38b4432ae91f3b148d80bacf482f1f48fb23b65e/lib/fluent/plugin/out_bigquery.rb#L27-L29

**Comment by sonots** [comment]
My thought: To support plugins such as fluent-plugin-bigquery, we have to change the behavior of `config_set_default` to change values of `conf`, or modify ways to pass parameters to internal plugins such as buffer, parser, formatter.

**Comment by sonots** [comment]
An easy way to fix is to change codes as:

```ruby
   class TimeSlicedOutput < BufferedOutput
     ...
     config_set_default :buffer_type, 'file'  # overwrite default buffer_type
-    config_set_default :buffer_chunk_limit, 256*1024*1024  # overwrite default buffer_chunk_limit
     config_set_default :flush_interval, nil

     def configure(conf)
 +     conf['buffer_chunk_limit'] ||= 256*1024*1024  # overwrite default buffer_chunk_limit
       super
```

and send similar PR to fluent-plugin-bigquery.

**Comment by sonots** [comment]
@tagomoris @repeatedly any thoughts?

**Comment by repeatedly** [comment]
Hmm... Put `conf['buffer_chunk_limit'] ||= 256*1024*1024` in the 3rd party plugin's configure seems enough for me.

**Comment by repeatedly** [comment]
Ah, wait. I misunderstood the issue.
TimeSlicedOutput code itself has this broken code.

**Comment by tagomoris** [comment]
This issue will be solved by this commit (only for v0.14).
https://github.com/fluent/fluentd/pull/866/commits/a446d6d5e6fd3c298920e82e939104c9aca339b5

**Comment by tagomoris** [comment]
The commit which solves this was merged into master.
